pkg/utils/tsoutil/tso_dispatcher.go

```diff
@@ -0,0 +1,220 @@
+	tsoDispatcher := &TSODispatcher{
+		tsoProxyHandleDuration: tsoProxyHandleDuration,
+		tsoProxyBatchSize:      tsoProxyBatchSize,
+	}
+	return tsoDispatcher
+}
+
+// DispatchRequest is the entry point for dispatching/forwarding a tso request to the detination host
+func (s *TSODispatcher) DispatchRequest(
+	ctx context.Context, req Request, tsoProtoFactory ProtoFactory, doneCh <-chan struct{}, errCh chan<- error) {
+	val, loaded := s.dispatchChs.LoadOrStore(req.getForwardedHost(), make(chan Request, maxMergeRequests))
+	reqCh := val.(chan Request)
+	if !loaded {
+		tsDeadlineCh := make(chan deadline, 1)
+		go s.dispatch(ctx, tsoProtoFactory, req.getForwardedHost(), req.getClientConn(), reqCh, tsDeadlineCh, doneCh, errCh)
+		go watchTSDeadline(ctx, tsDeadlineCh)
+	}
+	reqCh <- req
+}
+
+func (s *TSODispatcher) dispatch(
+	ctx context.Context, tsoProtoFactory ProtoFactory, forwardedHost string, clientConn *grpc.ClientConn,
+	tsoRequestCh <-chan Request, tsDeadlineCh chan<- deadline, doneCh <-chan struct{}, errCh chan<- error) {
+	dispatcherCtx, ctxCancel := context.WithCancel(ctx)
+	defer ctxCancel()
+	defer s.dispatchChs.Delete(forwardedHost)
+
+	log.Info("create tso forward stream", zap.String("forwarded-host", forwardedHost))
+	forwardStream, cancel, err := tsoProtoFactory.createForwardStream(ctx, clientConn)
+	if err != nil || forwardStream == nil {
+		log.Error("create tso forwarding stream error",
+			zap.String("forwarded-host", forwardedHost),
+			errs.ZapError(errs.ErrGRPCCreateStream, err))
+		select {
+		case <-dispatcherCtx.Done():
+			return
+		case _, ok := <-doneCh:
+			if !ok {
+				return
+			}
+		case errCh <- err:
+			close(errCh)
+			return
+		}
+	}
+	defer cancel()
+
+	requests := make([]Request, maxMergeRequests+1)
+	for {
+		select {
+		case first := <-tsoRequestCh:
+			pendingTSOReqCount := len(tsoRequestCh) + 1
+			requests[0] = first
+			for i := 1; i < pendingTSOReqCount; i++ {
+				requests[i] = <-tsoRequestCh
+			}
+			done := make(chan struct{})
+			dl := deadline{
+				timer:  time.After(DefaultTSOProxyTimeout),
+				done:   done,
+				cancel: cancel,
+			}
+			select {
+			case tsDeadlineCh <- dl:
+			case <-dispatcherCtx.Done():
+				return
+			}
+			err = s.processRequests(forwardStream, requests[:pendingTSOReqCount], tsoProtoFactory)
+			close(done)
+			if err != nil {
+				log.Error("proxy forward tso error",
+					zap.String("forwarded-host", forwardedHost),
+					errs.ZapError(errs.ErrGRPCSend, err))
+				select {
+				case <-dispatcherCtx.Done():
+					return
+				case _, ok := <-doneCh:
+					if !ok {
+						return
+					}
+				case errCh <- err:
+					close(errCh)
+					return
+				}
+			}
+		case <-dispatcherCtx.Done():
+			return
+		}
+	}
+}
+
+func (s *TSODispatcher) processRequests(forwardStream stream, requests []Request, tsoProtoFactory ProtoFactory) error {
+	// Merge the requests
+	count := uint32(0)
+	for _, request := range requests {
+		count += request.getCount()
+	}
+
+	start := time.Now()
```

Comment:
It previously appeared to be in a similar position in L143

Reply:
This is a better position for s.tsoProxyHandleDuration.Observe(time.Since(start).Seconds())